### ISSUE TITLE
Can ffprobe check for an English audio track?

### ISSUE THREAD

**Issue by neo-neo1** [clinton-hall/nzbToMedia#1853]
Can nzbToMedia have ffprobe check the download for an English audio track? And if it doesn't include English mark it as failed, so I can have failed downloaded handling automatically download another release which hopefully does. 

Quite common today, particularly on Netflix to have international based shows on Netflix USA available in multiple audio tracks (e.g. Squid Game). Unfortunately not all release groups include English audio tracks.

**Comment by clinton-hall** [comment]
Interesting...

If we had a parameter right below ` check_media = 1`
that specified the language to check for:
`require_lan = eng`
I would need to set a default to `all` so that this check is disabled by default until the user sets a language

then in
https://github.com/clinton-hall/nzbToMedia/blob/master/core/transcoder.py#L63-L73

Add some checks 
```
if video_details.get('streams'):
        video_streams = [item for item in video_details['streams'] if item['codec_type'] == 'video']
        audio_streams = [item for item in video_details['streams'] if item['codec_type'] == 'audio']
        if len(video_streams) > 0 and len(audio_streams) > 0 and (core.require_lan == 'all' or  core.require_lan in [item for item in audio_streams['tags']['language']):
            logger.info('SUCCESS: [{0}] has no corruption.'.format(file_name_ext), 'TRANSCODER')
            return True
        else:
            logger.info('FAILED: [{0}] has {1} video streams and {2} audio streams. '
                        'Assume corruption.'.format
                        (file_name_ext, len(video_streams), len(audio_streams)), 'TRANSCODER')
            return False
```
this in theory should return SUCCESS if the file has a video stream and an audio stream and (either the require_lan = 'all' or the require_lan is found in one of the audio streams 'language' tag...)

This is all theoretical. I need to get a test system set up and analyse this a bit closer.

**Comment by neo-neo1** [comment]
> Interesting...
> 
> If we had a parameter right below ` check_media = 1` that specified the language to check for: `require_lan = eng` I would need to set a default to `all` so that this check is disabled by default until the user sets a language
> 
> then in https://github.com/clinton-hall/nzbToMedia/blob/master/core/transcoder.py#L63-L73
> 
> Add some checks
> 
> ```
> if video_details.get('streams'):
>         video_streams = [item for item in video_details['streams'] if item['codec_type'] == 'video']
>         audio_streams = [item for item in video_details['streams'] if item['codec_type'] == 'audio']
>         if len(video_streams) > 0 and len(audio_streams) > 0 and (core.require_lan == 'all' or  core.require_lan in [item for item in audio_streams['tags']['language']):
>             logger.info('SUCCESS: [{0}] has no corruption.'.format(file_name_ext), 'TRANSCODER')
>             return True
>         else:
>             logger.info('FAILED: [{0}] has {1} video streams and {2} audio streams. '
>                         'Assume corruption.'.format
>                         (file_name_ext, len(video_streams), len(audio_streams)), 'TRANSCODER')
>             return False
> ```
> 
> this in theory should return SUCCESS if the file has a video stream and an audio stream and (either the require_lan = 'all' or the require_lan is found in one of the audio streams 'language' tag...)
> 
> This is all theoretical. I need to get a test system set up and analyse this a bit closer.

Damn you work quick! I look forward to you testing, please keep us updated. My current system is production, I can’t experiment on it. Thanks again!

**Comment by clinton-hall** [comment]
Yeah, I must admit, I have had to re-download a few things recently because the language was not identified in the download name... So this should help.

I believe I have the basic structure in place. I'll need to do some tests.
I have specifically added some additional tests to this since the way I first proposed here would have caused too many false-flag corruption reports back to the nzb host. i.e. I now try to determine the difference between a corrupt download and a wrong language download...

I also need to check that the language tag is consistent... i.e. if the tag can have `en` or `english` and we are checking for `eng` then this will cause a lot of issues.

**Comment by neo-neo1** [comment]
> 
> I also need to check that the language tag is consistent... i.e. if the tag can have `en` or `english` and we are checking for `eng` then this will cause a lot of issues.

Valid point. Perhaps a simple regex that looks for “en” “eng” or “english”?

**Comment by clinton-hall** [comment]
ok.... when the language tag is included, it appears to always be the 3 letter version ("eng", "spa", "fre" etc.).
I have fixed this to prevent errors when the media file doesn't include the language tags.... 

currently this is absolute. In that we ignore any vide that doesn't have the language (e.g. "eng"). So if for example you downloaded an avi (doesn't have tags) or otherwise download an .mp4 that doesn't have the language set in the audio stream(s) then this would ignore that file even if it does have an English audio track...

I am wondering about having an option to "reject" only "other" languages, but keep non-defined. Or otherwise ignore non-defined and keep only English etc...

**Comment by neo-neo1** [comment]
> I am wondering about having an option to "reject" only "other" languages, but keep non-defined. Or otherwise ignore non-defined and keep only English etc...

Perhaps the later seems optimal. Though I am not certain I fully comprehend your question. Either way, no plan survives contact with the enemy.

For[ language tags I found this information useful.](https://gitlab.com/mbunkus/mkvtoolnix/-/wikis/Languages-in-Matroska-and-MKVToolNix)

**Comment by clinton-hall** [comment]
Ok... I've tidied up the logging (debug) so users can see when videos are "failed" due to missing language vs corrupt.
I am now using the `media_lan_check` branch for my downloads. I should have a few over the next few days so hopefully I'll be able to confirm that this is all working as wanted. I'll then look to merge up to nightly branch and see if any users encounter issues with other configs/scenarios.

**Comment by clinton-hall** [comment]
ok... all looked good overnight. Now merged into nightly.

**Comment by clinton-hall** [comment]
ok... had a few shows download... I noticed 1 in particular failed 4 times, then a few hours later It grabbed a new release and all succeeded.

So I re-downloaded the failed releases to have a look. it turns out there was an early `MeGusta` and `SYNCOPY` release that both had NO language tags... they were in-fact English audio, but without building some sort of linguistics audio tool (and even then, how do you account for some shows that feature multiple languages, often at the start when they set up the 'bad guys'...?) there is no other way to determine the language.

The release I finally got was `GOSSIP` for what it's worth...

So. As it stands, this new feature rejected 4 releases (3 were actually re-posts of the same) that had no language tags in the audio tracks. These also had no language specified in the file or download name, and had no .nfo files to parse etc...
In short, I believe this has done what is asked. I did eventually get a download that was confirmed to be English.

Setting `require_lan` will likely cause more releases to be rejected, but ultimately you should be fairly confident that the release that is accepted will have the required language.
